### PR TITLE
Build dag

### LINKED ISSUE #41: Export workflow graph
**[Original report](https://bitbucket.org/glotzer/signac-flow/issue/41) by Carl Simon Adorf (Bitbucket: [csadorf](https://bitbucket.org/csadorf), GitHub: [csadorf](https://github.com/csadorf)).**

----------------------------------------

As per offline discussion, we determined that it should be possible to explicitly export the directed graph defined as `FlowProject`, *.e.g.*, as NetworkX graph structure.

In loose relation to that, we should think on how to submit a complete workflow instead of individual operations to a scheduler.

### PR DESCRIPTION
**Pull request by vyasr**
## Description
This PR implements operation graph detection. For testing this PR, the following FlowProject and testing script may be helpful (I had to make them text files for GitHub to allow the upload):
[project.py.txt](https://github.com/glotzerlab/signac-flow/files/3800728/project.py.txt)
[test_dag.py.txt](https://github.com/glotzerlab/signac-flow/files/3800730/test_dag.py.txt)

You can create an arbitrary random project and then run this script. I will create new tests based on these ideas (and update changelog etc) once we're happy with the implementation.

@csadorf I think this change should sufficiently address your concerns about the old PR. It uses simpler logic than the original version and is much more transparent. There is a little bit of additional complexity with the use of tags, but it should be easy to follow. The tags are necessary because the `__code__.co_code` of the automatically generated lambda functions that we use internally is unchanged by the argument, i.e. `FlowProject.pre.true('a')`, and `FlowProject.pre.true('b')` will evaluate to being identical. Unfortunately this would also affect similarly automatically generated user conditions, but we can document that users can provide the tag argument to condition construction if they wish to use the graph with such conditions.

## Motivation and Context

Supersedes #71, resolves #41.

## Types of Changes

- [ ] Documentation update
- [ ] Bug fix
- [x] New feature
- [ ] Breaking change<sup>1</sup>

<sup>1</sup>The change breaks (or has the potential to break) existing functionality.

## Checklist:

- [x] I am familiar with the [**Contributing Guidelines**](https://github.com/glotzerlab/signac-flow/blob/master/CONTRIBUTING.md).
- [x] I agree with the terms of the [**Contributor Agreement**](https://github.com/glotzerlab/signac-flow/blob/master/ContributorAgreement.md).
- [x] My name is on the [list of contributors](https://github.com/glotzerlab/signac-flow/blob/master/contributors.yaml).
- [x] My code follows the [code style guideline](https://github.com/glotzerlab/signac-flow/blob/master/CONTRIBUTING.md#code-style) of this project.
- [x] The changes introduced by this pull request are covered by existing or newly introduced tests.

If necessary:
- [x] I have updated the API documentation as part of the package doc-strings.
- [ ] I have created a separate pull request to update the [framework documentation](https://docs.signac.io/) on [signac-docs](https://github.com/glotzerlab/signac-docs) and linked it here.
- [x] I have updated the [changelog](https://github.com/glotzerlab/signac-flow/blob/master/changelog.txt).

### CODE REVIEW

File: flow/project.py [hunk 1]
```diff
@@ -86,48 +86,58 @@
 {template_vars}
 
 Filter functions can be used to format template variables in a specific way.
-For example: {{ project.get_id() | captialize }}.
+For example: {{ project.get_id() | capitalize }}.
```
Comment (csadorf) [inline flow/project.py]: This seems unrelated?

File: flow/project.py [hunk 2]
```diff
@@ -687,6 +702,89 @@ def label_func(func):
...
+
+        Given a FlowProject subclass defined in a module `project.py`, the
+        output graph could be visualized using Matplotlib and NetworkX with the
+        following code:
+
+        .. code-block:: python
+
+            import numpy as np
+            from matplotlib import pyplot as plt
+            import networkx as nx
```
Comment (csadorf) [inline flow/project.py]: I would swap the above two lines.

File: flow/project.py [hunk 3]
```diff
@@ -687,6 +702,89 @@ def label_func(func):
...
+
+            from project import Project
+
+            p = Project()
```
Comment (csadorf) [inline flow/project.py]: I'd prefer to not abbreviate `project` to `p` if this is possible without sacrificing readability.

File: flow/project.py [hunk 4]
```diff
@@ -687,6 +702,89 @@ def label_func(func):
...
+            ops = p.operations.keys()
+            adj = np.asarray(p.detect_operation_graph())
+
+            plt.figure()
+            g = nx.DiGraph(adj)
+            pos = nx.spring_layout(g)
+            nx.draw(g, pos)
+            nx.draw_networkx_labels(g, pos,
+                labels={key: name for (key, name) in zip(range(len(ops)), [o for o in ops])})
```
Comment (csadorf) [inline flow/project.py]: Maybe we can break this line?

File: flow/project.py [hunk 5]
```diff
@@ -687,6 +702,89 @@ def label_func(func):
...
+            plt.show()
+        """
+
+        def to_callbacks(conditions):
+            """Get the actual callables associated with FlowConditions."""
+            return [condition._callback for condition in conditions]
+
+        def unpack_conditions(condition_functions):
+            """Identify any metaconditions in the list and reduce them to the
+            functions they're composed of. The callbacks argument is used
```
Comment (csadorf) [inline flow/project.py]:
```suggestion
            functions that they are composed of. The callbacks argument is used
```

File: flow/project.py [hunk 6]
```diff
@@ -405,6 +415,18 @@ def __call__(self, job=None):
             return self._cmd.format(job=job)
 
 
+def _create_metacondition(condition_dict, *other_funcs):
```
Comment (csadorf) [inline flow/project.py]: _I had left the most important comment as part of my review on this function, but somehow it got lost during the submission? Anyways, I'll just write it again..._

Here we make an assumption here that all meta-conditions use the `all` function to aggregate multiple condition functions. I think this should be either made explicit in the function name or the aggregation function (`all`, `any`, etc.) should be an argument.

File: flow/project.py (hunk 1 above)
Comment (vyasr) [reply flow/project.py]: It is, but it's such a minor typo fix that I was willing to put it into this PR.

File: flow/project.py (hunk 6 above)
Comment (vyasr) [reply flow/project.py]: This is a good point. For now I'm going to just make it explicit in the name because I'm not exactly certain how the `composed_of` logic for graphs would work for other aggregation functions and I don't see a purpose to generalizing until we have an actual use-case.

**Comment by mikemhenry** [comment]
> I had to make them text files for GitHub to allow the upload

What year is it github?! That annoys me so much. I can take a look at this tonight or tomorrow.

**Comment by vyasr** [comment]
@mikemhenry sounds good. I recall that you reviewed an earlier PR on this topic so I figured you'd be a good second reviewer (I think @bdice has had some other PR reviews requested recently so I moved it to you).

File: flow/project.py (hunk 6 above)
Comment (csadorf) [reply flow/project.py]: I think it's fine to move forward with this without much further consideration since this is purely internal anyways.

File: flow/project.py (hunk 1 above)
Comment (csadorf) [reply flow/project.py]: According to our [governance](https://signac.io/governance/) 

> Committers may commit small changes (documentation updates, code comments, CI configuration changes) without a pull request

I think this falls under this.

The reason that I am so anal about this particular rule, is because we generally squash all commits prior to merging PRs and that means that these unrelated changes are now convoluted with the actually relevant changes. This is highly confusing to anybody who is reading the git log and attempts to understand the motivation for a particular change.

Comment (mikemhenry) [reply flow/project.py]: I'm with @csadorf on this one, I can't find the article now but if we are strict about things like this, git blame and other git features work best when we keep the git log/commit history concise and "right"

File: flow/project.py [hunk 7]
```diff
@@ -86,48 +86,58 @@
...
+For example: {{ project.get_id() | capitalize }}.
 
 The available filters are:
 {filters}"""
 
 
 class _condition(object):
-
-    def __init__(self, condition):
+    # Random tag id used to ensure that "always" and "never" conditions never
```
Comment (mikemhenry) [inline flow/project.py]: This is more of a clarifying question, but I don't see how the the tag id is 'random'. I think it either needs a better name (arbitrary?) or the comment here needs to explain.

File: flow/project.py (hunk 1 above)
Comment (vyasr) [reply flow/project.py]: Fair enough, I can revert this change then.

File: flow/project.py (hunk 7 above)
Comment (vyasr) [reply flow/project.py]: You're correct, arbitrary is a more appropriate choice of words. I'll change that name.

Comment (vyasr) [reply flow/project.py]: For clarification, it only exists to ensure that these conditions (always and never) are not found to be equal.

**Comment by vyasr** [comment]
@mikemhenry I think I've addressed your question now. Feel free to merge if you think it's ready.
